### PR TITLE
Datatype codec support

### LINKED ISSUE #11: Implement all the data type mappings in the TCK

### PR DESCRIPTION
**Pull request by BillyYccc**
fixes #11

### CODE REVIEW

File: src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java [hunk 1]
```diff
@@ -39,18 +50,137 @@ static String inferenceParamDefinitionByValueType(Object value) {
...
+
+  private static LocalDate decodeDateN(ByteBuf in) {
+    byte dateLength = in.readByte();
+    if (dateLength == 0) {
+      return null;
+    } else if (dateLength == 3) {
+      int days = in.readUnsignedMediumLE();
+      return START_DATE.plus(days, ChronoUnit.DAYS);
+    } else {
+      throw new IllegalStateException();
```
Comment (vietj) [inline src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java]: Add error message like "Unexpected encoded date length"

Comment (vietj) [reply src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java]: also give the value that was read for troubleshooting

File: src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java [hunk 2]
```diff
@@ -39,18 +50,137 @@ static String inferenceParamDefinitionByValueType(Object value) {
...
+        case 4:
+          value = in.readIntLE();
+          break;
+        case 8:
+          value = in.readLongLE();
+          break;
+        case 12:
+        case 16:
+          //TODO wrapped types?
+          throw new UnsupportedOperationException();
```
Comment (vietj) [inline src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java]: Do we need a test for this ?

**Review by vietj — COMMENTED** [review]
Looking good, however sometimes `IllegalStateException` is thrown which is fine for protocol errors however it should add some context like the unexpected value to help finding out the actual issue

File: src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java (hunk 2 above)
Comment (BillyYccc) [reply src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java]: updated with https://github.com/BillyYccc/incubator-vertx-mssql-client/pull/18/commits/03f2b1be16911f40974a84e4a6206048437d06e0.

File: src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java (hunk 1 above)
Comment (BillyYccc) [reply src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java]: fixed by https://github.com/BillyYccc/incubator-vertx-mssql-client/pull/18/commits/4b40eb52219c93a9985071939f583294b33f6747.

**Comment by BillyYccc** [comment]
I have fixed the issues according to the review, can you please see it?

**Review by vietj — APPROVED** [review]

**Comment by BillyYccc** [comment]
Thanks for the review!

**Pull request merged**